src/metabase/serialization/dump.clj

```diff
@@ -0,0 +1,294 @@
+    :segment       [op (fully-qualified-name (Segment (first args)))]
+    :field-id      [op (if (string? (first args))
+                         (first args)
+                         (fully-qualified-name (Field (first args))))]
+    :fk->          (into [op]
+                         (for [arg args]
+                           (if (number? arg)
+                             (Field arg)
+                             (entity-reference->fully-qualified-name arg))))
+    :field-literal entity-reference))
+
+(defn- humanize-entity-references
+  [entity]
+  (walk/postwalk
+   (fn [form]
+     (cond
+       (entity-reference? form)
+       (entity-reference->fully-qualified-name form)
+
+       (map? form)
+       (let [id->fully-qualified-name (fn [entity-id model]
+                                        (if (string? entity-id)
+                                          entity-id
+                                          (fully-qualified-name (model entity-id))))]
+         (-> form
+             (u/update-when :database (fn [db]
+                                        (if (= db -1337)
+                                          "database/virtual"
+                                          (id->fully-qualified-name db Database))))
+             (u/update-when :card_id id->fully-qualified-name Card)
+             (u/update-when :source-table (fn [source-table]
+                                            (if (and (string? source-table)
+                                                     (str/starts-with? source-table "card__"))
+                                              (-> source-table
+                                                  (str/split #"__")
+                                                  second
+                                                  Integer/parseInt
+                                                  (id->fully-qualified-name Card))
+                                              (id->fully-qualified-name source-table Table))))))
+
+       :else
+       form))
+   entity))
+
+(defmulti
+  ^{:doc      "Serialize entity `entity` to location `dir`.
+
+               Depending on the entity, it will be serialized either as a single YAML, or a
+               directory structure.
+
+               Removes unneeded fields that can either be reconstructed from context or are
+               meaningless (eg. :created_at)."
+    :arglists '([dir entity])}
+  dump (fn [_ entity]
+         (type entity)))
+
+(defn- strip-crud
+  [entity]
+  (cond-> (dissoc entity :id :creator_id :created_at :updated_at :db_id :database_id :location
+                  :dashboard_id :fields_hash :personal_owner_id :made_public_by_id :collection_id
+                  :pulse_id)
+    (some #(instance? % entity) (map type [Metric Field Segment])) (dissoc :table_id)))
+
+(defn- spit-yaml
+  [fname obj]
+  (io/make-parents fname)
+  (spit fname (yaml/generate-string obj :dumper-options {:flow-style :block})))
```

Comment:
Personal option, but I think this namespace could benefit from narrowing its focus and moving the YAML related stuff into a different `dump.yaml` namespace or something like that. 

It would be good to have one namespace that focuses on the underlying logic of generating dumps and creating a Clojure representation of them and have another namespace take that dump and worry about writing it to YAML. Or maybe some configuration even more focused than that. Division of labor / clear responsibilities /small, focused modules and all that. Would be easier to test that way. And if we wanted to do something wacky like support EDN as a dump format it that code wouldn't have to go in one uber-namespace as well. (Not saying we want to -- I don't -- but I think in this case organizing things in an extensible way also means they'd be organized in a clearer way.)

Reply:
I guess to do that you'd want to move the logic to determine path names into that other namespace; since that uses `fully-qualified-name` as well maybe that method and its impls could go in a shared `dump.names` namespace or something.

Either way, I think serializing an object is fundamentally a separate operation from writing that serialized object to YAML -- there's a real possibility we'll want to be able to do one without the other at some point in the future. Perhaps in the future Metabase plugins can add a new H2 or SQLite database and define Cards and Dashboards for it using the same deserialization/serialization logic here -- we could package up additional sample datasets that way. It would be pretty cool to be able to click a few buttons and have a new test dataset to play around with.

Thus I think you should really reconsider tying together serializing an object and writing that object to YAML -- ideally, those would be two different operations.

Perhaps a namespace architecture and general flow like this could work?
```
;;                                         serialization
;;                public interface, Calls methods in serialize to serialize objects;
;;                then passes the serialized versions to YAML backend which writes them to YAML
;;                        1                 ↓ ↑     ↓
;;          +-------------------------------+ |     |
;;          |             objects             |     |
;;          ↓                   serialized  4 |     | 5
;; serialization.serialize -------------------+     | serialized
;; methods/impls for converting objects             |
;; to serializable forms                            |
;;      |         ↑                                 |
;;    2 |         | 3                               |
;;      ↓         |                                 ↓         10
;; serialization.names                    serialization.yaml ---> YAML files
;; fully-qualified-name method & impls         |       ↑
;;       ↓      ↑                            6 |       | 9
;;       |      |                              |       |
;;       |      |                              ↓       |
;;       |      |                         serialization.paths
;;       |      |                   logic (potentially) shared between different backends to
;;       |      |                   determine paths where things should be written to. Uses
;;       |      |                   logic from `serializaition.names`
;;       |      |               7              ↓       ↑
;;       |      +------------------------------+       |
;;       +---------------------------------------------+
;;                             8        
```

Reply:
@sbelak were you still planning on shuffling things around like this? Separating the concept of serializing objects from writing them to YAML